zmon-controller-app/src/main/java/org/zalando/zmon/service/impl/Grafana.java

```diff
@@ -0,0 +1,156 @@
+                .queryParam("orgId", "1")
+                .queryParam("checkId", params.containsKey("checkId") ? params.get("checkId") : "")
+                .queryParam("entityName", params.containsKey("entityName") ? params.get("entityName") : "")
+                .queryParam("checkName", params.containsKey("checkName") ? params.get("checkName") : "")
+                .queryParam("refresh", "1m")
+                .build();
+
+        return "redirect:" + uri.toUriString();
+    }
+
+    @Override
+    public ResponseEntity<JsonNode> getDashboard(String uid, String token) {
+        final Executor executor = Executor.newInstance(visualizationProperties.getHttpClient());
+        final String url = visualizationProperties.getUrl() + getAndDeleteDashboardEndPoint + uid;
+
+        try {
+            Request request = Request.Get(url);
+            request.addHeader("Authorization", "Bearer " + token);
+            HttpResponse response = executor.execute(request).returnResponse();
+            return toResponseEntity(response);
+        } catch (Exception ex) {
+            log.error("Get grafana dashboard: {} failed", uid, ex);
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @Override
+    public ResponseEntity<JsonNode> searchDashboards(Map<String, String> params, String token) {
+        final Executor executor = Executor.newInstance(visualizationProperties.getHttpClient());
+
+        try {
+            log.info("Searching grafana dashboard: Query={} User={}",
+                    params.containsKey("query") ? URLEncoder.encode(params.get("query"), "UTF-8") : "",
+                    authService.getUserName());
+            UriComponents url = UriComponentsBuilder.fromUriString(visualizationProperties.getUrl())
+                    .path(searchDashboardEndpoint)
+                    .queryParam("query", params.containsKey("query") ? URLEncoder.encode(params.get("query"), "UTF-8") : "")
+                    .queryParam("tag", params.containsKey("tag") ? URLEncoder.encode(params.get("tag"), "UTF-8") : "")
+                    .queryParam("limit", params.containsKey("limit") ? params.get("limit") : "25")
+                    .build();
+            Request request = Request.Get(url.toUri());
+            request.addHeader("Authorization", "Bearer " + token);
+            HttpResponse response = executor.execute(request).returnResponse();
+            return toResponseEntity(response);
+        } catch (Exception ex) {
+            log.error("Search grafana dashboard failed", ex);
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @Override
+    public ResponseEntity<JsonNode> upsertDashboard(String dashboard, String token) {
+        log.info("Creating/Updating grafana dashboard: user={}", authService.getUserName());
+        final Executor executor = Executor.newInstance(visualizationProperties.getHttpClient());
```

Comment:
Are you sure we have to create `Executor` on every call? Usually they are stateless, so can be created in constructor and reused between calls